### PR TITLE
WIP: Harmonize calls to EVP_Digest*{Init,Update,Final}()

### PR DESCRIPTION
**Pull request by mspncp**
Commit 2 was motivated by @Bugcheckers remark https://github.com/openssl/openssl/pull/6819#issuecomment-410296964. The other two commits are by-products.

### Commit 1

```
    EVP_DigestInit.pod: add missing entries to RETURN VALUES section
```

### Commit 2

```
    Harmonize calls to EVP_Digest*Update()
    
    The majority of callers does not check for an empty buffer (buffer_length == 0) before calling
    EVP_Digest*Update(), because all implementations handle this case correctly as no-op.
    (See discussions in #6800, #6817, and #6838.)
```

### Commit 3

Noticed while working on commit 2. Can be squashed with commit 2 if desired, or discarded if you think it is consistency overkill.

```
    Harmonize calls to EVP_Digest*{Init,Update,Final}()
    
    According to the manual pages, the functions return a boolean value,
    so checking for `EVP_Digest*() <= 0` is misleading. Also the
    statistics show a clear tendency towards using `!EVP_*()`
    
    ~/src/openssl$ grep -rn 'EVP_Digest\(Sign\)\?\(Init\|Update\|Final\).*<=' | wc -l
    77
    ~/src/openssl$ grep -rn '!EVP_Digest\(Sign\)\?\(Init\|Update\|Final\)' | wc -l
    245
```

### CODE REVIEW

**Comment by kroeckx** [comment]
At least this test fails:
```
        not ok 1 - ../../../_srcdist/test/recipes/30-test_evp_data/evppkey.txt
    not ok 1 - run_file_tests
../../util/shlib_wrap.sh ../../test/evp_test ../../../_srcdist/test/recipes/30-test_evp_data/evppkey.txt => 1
not ok 7 - running evp_test evppkey.txt
```

File: crypto/evp/m_sigver.c [hunk 1]
```diff
@@ -158,7 +158,7 @@ int EVP_DigestSign(EVP_MD_CTX *ctx, unsigned char *sigret, size_t *siglen,
 {
     if (ctx->pctx->pmeth->digestsign != NULL)
         return ctx->pctx->pmeth->digestsign(ctx, sigret, siglen, tbs, tbslen);
-    if (sigret != NULL && EVP_DigestSignUpdate(ctx, tbs, tbslen) <= 0)
```
[the remark this answers is not in the document]
Comment (levitte) [reply crypto/evp/m_sigver.c]: No. If `sigret == NULL`, then it's expected that the caller tries to figure out what the signature length will be. In that case, the `EVP_DigestSignUpdate` call is a waste of cycles (and I have no idea what corner cases it might hit as well)

**Comment by mspncp** [comment]
Thanks Richard! The latest fixup seems heal most of the evp_test, except for the SM2 tests, which I will have to look at later. Strangely, the test outcome is 'ok' although the SM2 tests report a mismatch. Is this a bug in the test?


```
    # Starting "SM2 tests" tests at line 18402
    # ERROR: (memory) 'expected->output == got' failed @ test/evp_test.c:1111
    # --- expected->output
    # +++ got
    # 0000:-3045022100f11bf3 6e75bb304f094fb4 2a4ca22377d0cc76 8637c5011cd59fb9
    # 0000:+3045022061fa912d d72fff1aebf4857a 77189bdf25c2ae63 fe6d530e7878d89d
    #             ^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-ed4b130c98022035 545ffe2c2efb3abe e4fee661468946d8 86004fae8ea53115
    # 0020:+570505b602210099 be0398f934a02cc6 d47cf774e44294ad b35c6741ad81693c
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0040:-93e48f7fe21b91
    # 0040:+a14caef35d5af2
    #       ^^^^^^^^^^^^^^
    #
    # INFO:  @ test/testutil/stanza.c:121
    # Starting "Chosen Wycheproof vectors" tests at line 18440
    # INFO:  @ test/testutil/stanza.c:33
    # Completed 1442 tests with 0 errors and 0 skipped
    ok 1 - test/recipes/30-test_evp_data/evppkey.txt
ok 1 - run_file_tests
```

**Comment by mspncp** [comment]
I created issue #6857 for the test problem.

**Comment by mspncp** [comment]
Since @romen and @kroeckx pointed out to me that the SM2 test was not really failing, I think this pull request is ready for review now. I rebased, squashed the fixup and reordered the commits but did not make any additional source code changes. Only a little rewording of the commit messages.

File: ssl/ssl_lib.c [hunk 2]
```diff
@@ -4404,7 +4404,7 @@ EVP_MD_CTX *ssl_replace_hash(EVP_MD_CTX **hash, const EVP_MD *md)
 {
     ssl_clear_hash_ctx(hash);
     *hash = EVP_MD_CTX_new();
-    if (*hash == NULL || (md && EVP_DigestInit_ex(*hash, md, NULL) <= 0)) {
+    if (*hash == NULL || (md && !EVP_DigestInit_ex(*hash, md, NULL))) {
```
Comment (levitte) [inline ssl/ssl_lib.c]: Hmmm... Do we call this function with a NULL md, ever? Is it reasonable to expect it to be NULL? If not, then I don't see why we're testing it...

Comment (mspncp) [reply ssl/ssl_lib.c]: Hmmm, good question... Its seems to be the only place in the entire source code where this is checked. 
```
~/src/openssl$ rgrep EVP_DigestInit | grep -F '&&'
ssl/ssl_lib.c:4407:    if (*hash == NULL || (md && !EVP_DigestInit_ex(*hash, md, NULL))) {
```
Probably you're right. I will remove the check and see what the tests say about it.

Comment (mspncp) [reply ssl/ssl_lib.c]: Just noticed right now that the [doc comment]( https://github.com/mspncp/openssl/blob/226e65c417e234d5cdc6a5b19b7c5a2197d11920/ssl/ssl_lib.c#L4401) says that the md is optional. So even if nobody would use this feature we still couldn't change it in this release.

**Comment by mspncp** [comment]
ping

**Comment by mspncp** [comment]
ping @openssl/omc for a second review.

**Comment by levitte** [comment]
Speaking of harmonization, have you read `doc/man3/EVP_DigestSignInit.pod`?  Take an extra look at the "RETURN VALUES" section.  That would need an update, no?

**Comment by mspncp** [comment]
Hmmm, ok...   thanks for pointing out that inconsistency with the documentation. I will have to check the sources first whether the documentation is correct and `EVP_DigestSign*()` and `EVP_Digest*()` indeed return different kind of values. The `EVP_DigestSign*()` call statistics is  undecided; either way, approximately half of the calls could be harmonized.

**EVP_DigestSign**
```
~/src/openssl$ grep -rn 'EVP_DigestSign\(Init\|Update\|Final\).*<=' | wc -l
26
~/src/openssl$ grep -rn '!EVP_DigestSign\(Init\|Update\|Final\)' | wc -l
20
```
**EVP_Digest**
```
~/src/openssl$ grep -rn 'EVP_Digest\(Init\|Update\|Final\).*<=' | wc -l
51
~/src/openssl$ grep -rn '!EVP_Digest\(Init\|Update\|Final\)' | wc -l
225
```

**Comment by mspncp** [comment]
This pull request has aged and I currently have no plans to resume it.
